internal/peermanager/peerwantmanager.go

```diff
@@ -176,57 +182,45 @@ func (pwm *peerWantManager) prepareSendWants(p peer.ID, wantBlocks []cid.Cid, wa
 			pwm.reverseIndexAdd(c, p)
 
 			// Add the CID to the results
-			resWantHvs = append(resWantHvs, c)
+			fltWantHvs = append(fltWantHvs, c)
 		}
 	}
 
-	return resWantBlks, resWantHvs
+	// Send the want-blocks and want-haves to the peer
+	pws.peerQueue.AddWants(fltWantBlks, fltWantHvs)
 }
 
-// PrepareSendCancels filters the list of cancels for each peer,
-// returning a map of peers which only contains cancels for wants that have
-// been sent to the peer.
-func (pwm *peerWantManager) prepareSendCancels(cancelKs []cid.Cid) map[peer.ID][]cid.Cid {
+// sendCancels sends a cancel to each peer to which a corresponding want was
+// sent
+func (pwm *peerWantManager) sendCancels(cancelKs []cid.Cid) {
 	if len(cancelKs) == 0 {
-		return nil
-	}
-
-	// Pre-allocate enough space for all peers that have the first CID.
-	// Chances are these peers are related.
-	expectedResSize := 0
-	firstCancel := cancelKs[0]
-	if pwm.broadcastWants.Has(firstCancel) {
-		expectedResSize = len(pwm.peerWants)
-	} else {
-		expectedResSize = len(pwm.wantPeers[firstCancel])
+		return
 	}
-	res := make(map[peer.ID][]cid.Cid, expectedResSize)
 
-	// Keep the broadcast keys separate. This lets us batch-process them at
-	// the end.
+	// Handle broadcast wants up-front
 	broadcastKs := make([]cid.Cid, 0, len(cancelKs))
-
-	// Iterate over all requested cancels
 	for _, c := range cancelKs {
-		// Handle broadcast wants up-front.
-		isBroadcast := pwm.broadcastWants.Has(c)
-		if isBroadcast {
+		if pwm.broadcastWants.Has(c) {
 			broadcastKs = append(broadcastKs, c)
 			pwm.broadcastWants.Remove(c)
 		}
+	}
 
-		// Even if this is a broadcast, we may have sent targeted wants.
-		// Deal with them.
-		for p := range pwm.wantPeers[c] {
-			pws, ok := pwm.peerWants[p]
-			if !ok {
-				// Should never happen but check just in case
-				log.Errorf("peerWantManager reverse index missing peer %s for key %s", p, c)
+	// Allocate a single buffer to filter the cancels to send to each peer
+	cancelsBuff := make([]cid.Cid, 0, len(cancelKs))
+
+	// Send cancels to a particular peer
+	send := func(p peer.ID, pws *peerWant) {
+		// Include broadcast cancels
+		peerCancels := append(cancelsBuff[:0], broadcastKs...)
```

Comment:
nit: you don't _have_ to copy these every time.

Reply:
You're right, good idea 💡 I updated the benchmark stats above